clippy_lints/src/infinite_iter.rs

```diff
@@ -0,0 +1,218 @@
+use rustc::hir::*;
+use rustc::lint::*;
+use utils::{get_trait_def_id, implements_trait, higher, match_qpath, paths, span_lint};
+
+/// **What it does:** Checks for iteration that is guaranteed to be infinite.
+///
+/// **Why is this bad?** While there may be places where this is acceptable
+/// (e.g. in event streams), in most cases this is simply an error.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// repeat(1_u8).iter().collect::<Vec<_>>()
+/// ```
+declare_lint! {
+    pub INFINITE_ITER,
+    Warn,
+    "infinite iteration"
+}
+
+/// **What it does:** Checks for iteration that may be infinite.
+///
+/// **Why is this bad?** While there may be places where this is acceptable
+/// (e.g. in event streams), in most cases this is simply an error.
+///
+/// **Known problems:** The code may have a condition to stop iteration, but
+/// this lint is not clever enough to analyze it.
+///
+/// **Example:**
+/// ```rust
+/// [0..].iter().zip(infinite_iter.take_while(|x| x > 5))
+/// ```
+declare_lint! {
+    pub MAYBE_INFINITE_ITER,
+    Allow,
+    "possible infinite iteration"
+}
+
+#[derive(Copy, Clone)]
+pub struct Pass;
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INFINITE_ITER, MAYBE_INFINITE_ITER)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        let (lint, msg) = match complete_infinite_iter(cx, expr) {
+            True => (INFINITE_ITER, "infinite iteration detected"),
+            Unknown => (MAYBE_INFINITE_ITER,
+                        "possible infinite iteration detected"),
+            False => { return; }
+        };
+        span_lint(cx, lint, expr.span, msg)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum TriState {
+    True,
+    Unknown,
+    False
+}
+
+use self::TriState::{True, Unknown, False};
+
+impl TriState {
+    fn and(self, b: Self) -> Self {
+        match (self, b) {
+            (False, _) | (_, False) => False,
+            (Unknown, _) | (_, Unknown) => Unknown,
+            _ => True
+        }
+    }
+
+    fn or(self, b: Self) -> Self {
+        match (self, b) {
+            (True, _) | (_, True) => True,
+            (Unknown, _) | (_, Unknown) => Unknown,
+            _ => False
+        }
+    }
+}
+
+impl From<bool> for TriState {
+    fn from(b: bool) -> Self {
+        if b { True } else { False }
+    }
+}
+
+#[derive(Copy, Clone)]
+enum Heuristic {
+    Always,
+    First,
+    Any,
+    All
+}
+
+use self::Heuristic::{Always, First, Any, All};
+
+// here we use the `TriState` as (Finite, Possible Infinite, Infinite)
```

Comment:
As this is the only use case for TriState, I'd rename it to Finitude and the variants to Finite, PossiblyInfinite, and Infinite.

Reply:
I had that initially, but renamed it to be more general when I added the second function, which would have used yet another tristate before I simplified the whole thing.

Perhaps we'll use a general TriState in the future, but for now I'll rename.